### PR TITLE
feat(query): migrate comp

### LINKED ISSUE #7636: Migrate comparison functions to new expression framework
**Summary**

Description for this feature.

### PR DESCRIPTION
**Pull request by TCeason**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

migrate func v1 comp to function v2

Fixes #7636

### CODE REVIEW

File: src/query/functions-v2/src/scalars/comparison.rs [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+use common_expression::types::StringType;
+use common_expression::types::ALL_NUMERICS_TYPES;
+use common_expression::with_number_mapped_type;
+use common_expression::FunctionProperty;
+use common_expression::FunctionRegistry;
+use regex::bytes::Regex as BytesRegex;
+use regex::bytes::RegexBuilder as BytesRegexBuilder;
+
+pub fn register(registry: &mut FunctionRegistry) {
+    registry.register_2_arg::<StringType, StringType, BooleanType, _, _>(
```
Comment (sundy-li) [inline src/query/functions-v2/src/scalars/comparison.rs:29]: BooleanType, TimestampType are not covered.

Can use macro to generate the codes like `ALL_NUMERICS_TYPES`

Comment (TCeason) [reply src/query/functions-v2/src/scalars/comparison.rs:29]: Got

File: src/query/functions-v2/src/scalars/comparison.rs [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+                        |lhs, rhs| lhs > rhs,
+                    );
+                    }
+                });
+            }
+            _ => todo!(),
+        });
+        with_cmp_mapped_type!(|DATA_TYPE| match ty {
+            DataType::DATA_TYPE => {
+                registry.register_2_arg::<DATA_TYPE, DATA_TYPE, BooleanType, _, _>(
```
Comment (sundy-li) [inline src/query/functions-v2/src/scalars/comparison.rs:159]: can we move these code together with "eq",  "neq", "gt"  ?

File: src/query/functions-v2/src/scalars/comparison.rs [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+    //     |_lhs, _rhs| None,
+    //     |lhs, rhs| lhs.cmp(rhs) != Ordering::Equal,
+    // );
+
+    registry.register_2_arg::<StringType, StringType, BooleanType, _, _>(
+        "like",
+        FunctionProperty::default(),
+        |_lhs, _rhs| None,
+        |lhs, rhs| {
+            let pattern_str = simdutf8::basic::from_utf8(rhs)
```
Comment (sundy-li) [inline src/query/functions-v2/src/scalars/comparison.rs:312]: need to cache the pattern regexp, we can add todo comments there and improve it in the next pr.

Comment (TCeason) [reply src/query/functions-v2/src/scalars/comparison.rs:245]: hmm, maybe we can create a hashmap to store the pattern?

File: src/query/functions-v2/src/scalars/comparison.rs [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+    // );
+
+    // registry.register_2_arg::<VariantType, VariantType, BooleanType, _, _>(
+    //     "noteq",
+    //     FunctionProperty::default(),
+    //     |_lhs, _rhs| None,
+    //     |lhs, rhs, _| lhs.cmp(rhs) != Ordering::Equal,
+    // );
+
+    registry.register_2_arg::<StringType, StringType, BooleanType, _, _>(
```
Comment (b41sh) [inline src/query/functions-v2/src/scalars/comparison.rs:240]: do we need to consider using `register_passthrough_nullable_2_arg` for nullable data types?

Comment (TCeason) [reply src/query/functions-v2/src/scalars/comparison.rs]: Yes, I try to modify this. But it occurs other err, needs to fix it next.

Comment (b41sh) [reply src/query/functions-v2/src/scalars/comparison.rs]: sorry, `register_2_arg` is actually a wrapper of `register_passthrough_nullable_2_arg`. Don't need to modify it.

File: src/query/functions-v2/src/scalars/comparison.rs [hunk 5]
```diff
@@ -0,0 +1,456 @@
...
+    //     |_lhs, _rhs| None,
+    //     |lhs, rhs, _| lhs.cmp(rhs) != Ordering::Equal,
+    // );
+
+    registry.register_passthrough_nullable_2_arg::<StringType, StringType, BooleanType, _, _>(
+        "like",
+        FunctionProperty::default(),
+        |_lhs, _rhs| None,
+        |lhs, rhs, ctx| {
+            let mut map: HashMap<&[u8], Regex> = HashMap::new();
```
Comment (TCeason) [inline src/query/functions-v2/src/scalars/comparison.rs:251]: Try to add the map into functioncontext but the struct functioncontext derived Copy trait. 

HashMap<&[u8], Regex> the Regex do not derived Copy.

So I can add it into functioncontext..

Try to use a closer function but the &[u8] has not enough life cycle...

So use this way to catch the map.

Is there has some way to optimize the function? I think I need some help.

Comment (b41sh) [reply src/query/functions-v2/src/scalars/comparison.rs]: how about use `register_function_factory`? like `regexp_instr_fn`

Comment (TCeason) [reply src/query/functions-v2/src/scalars/comparison.rs]: > how about use `register_function_factory`? like `regexp_instr_fn`

Yes, that's also a good idea. And I talk with andy, he has another way to resolve this.

File: src/query/functions-v2/tests/it/scalars/testdata/comparison.txt [hunk 6]
```diff
@@ -0,0 +1,1030 @@
...
+ast            : 1.1=1.1
+raw expr       : eq(1.1_f64, 1.1_f64)
+checked expr   : eq<Float64, Float64>(1.1_f64, 1.1_f64)
+optimized expr : true
+output type    : Boolean
+output domain  : Unknown
+output         : true
+
+
+ast            : parse_json(lhs) = parse_json(rhs)
```
Comment (TCeason) [inline src/query/functions-v2/tests/it/scalars/testdata/comparison.txt:113]: Please review variant CMP result is ok. @b41sh

Comment (TCeason) [reply src/query/functions-v2/tests/it/scalars/testdata/comparison.txt]: narrowed.

File: src/query/functions-v2/tests/it/scalars/comparison.rs [hunk 7]
```diff
@@ -0,0 +1,451 @@
...
+                r#""\\\"abc\\\"""#,
+                r#""databend""#,
+                r#"{"k":"v","a":"d"}"#,
+                r#"[1,2,3,["a","b","c"]]"#,
+            ]),
+        ),
+    ];
+    run_ast(
+        file,
+        r#"parse_json('[1,2,3,["a","b","c"]]') = parse_json('[1,2,3,["a","b","c"]]')"#,
```
Comment (b41sh) [inline src/query/functions-v2/tests/it/scalars/comparison.rs:132]: `parse_json(lhs) = parse_json(rhs)` ?

**Review by b41sh — APPROVED** [review]

**Review by sundy-li — APPROVED** [review]

**Pull request merged**